src/domain/mongoose_domain_api.erl

```diff
@@ -0,0 +1,136 @@
+
+-export([init/2,
+         insert_domain/2,
+         remove_domain/2,
+         disable_domain/1,
+         enable_domain/1,
+         get_host_type/1,
+         get_all_locked/0,
+         get_domains_by_host_type/1]).
+
+-type domain() :: jid:lserver().
+-type host_type() :: binary().
+-type pair() :: {domain(), host_type()}.
+
+%% Init - on init all the “enabled” domain names from the persistent storage
+%% must be added to the core MIM component described above.
+%% Domains should be nameprepped using `jid:nameprep'
+-spec init([pair()], [host_type()]) -> ok | {error, term()}.
+init(Pairs, AllowedHostTypes) ->
+    mongoose_domain_core:start(Pairs, AllowedHostTypes).
+
+%% Add domain name (w/ host type) - This function must be idempotent.
+%% Added domain is always “enabled” by default it must be added in the core
+%% MIM component described in the previous section.
+%% If it’s successfully enabled than Information about the domain name
+%% must be added into persistent storage and distributed across all the nodes
+%% in the cluster.
+%% Domain should be nameprepped using `jid:nameprep'.
+%% HostType should be opaque binary.
+-spec insert_domain(domain(), host_type()) ->
+    ok  | {error, duplicate} | {error, {db_error, term()}}
+    | {error, service_disabled} | {error, unknown_host_type}.
+insert_domain(Domain, HostType) ->
+    case check_domain(Domain, HostType) of
+        ok ->
+            check_db(mongoose_domain_sql:insert_domain(Domain, HostType));
+        Other ->
+            Other
+    end.
+
+%% Removal the domain name - This function must be idempotent.
+%% domain name must be removed from the core MIM component (if required)
+%% and from the DB. this action must be distributed across
+%% all the nodes in the cluster.
+%% Returns ok, if domain not found.
+%% Domain should be nameprepped using `jid:nameprep'.
+-spec remove_domain(domain(), host_type()) ->
+    ok | {error, locked} | {error, {db_error, term()}}
+    | {error, service_disabled} | {error, wrong_host_type} | {error, unknown_host_type}.
+remove_domain(Domain, HostType) ->
+    case check_domain(Domain, HostType) of
+        ok ->
+            check_db(mongoose_domain_sql:remove_domain(Domain, HostType));
+        Other ->
+            Other
+    end.
+
+%% Disabling/Enabling domain name - This function must be idempotent.
+%% the status of the existing domain must be changed.
+%% If domain name is enabled, then it must be added in the core MIM component.
+%% On disabling domain name must be removed from the core MIM component.
+%% Change of the status must be distributed across all the nodes in the cluster.
+-spec disable_domain(domain()) ->
+    ok | {error, not_found} | {error, locked} | {error, duplicate}
+    | {error, service_disabled} | {error, unknown_host_type}.
+disable_domain(Domain) ->
+    case mongoose_domain_core:is_locked(Domain) of
+        true ->
+            {error, locked};
+        false ->
+            case service_domain_db:enabled() of
```

Comment:
This check seems prone to race conditions, why not just check if the service is enabled in the config?

Reply:
changed